src/NITTA/Model/Microarchitecture/Builder.hs

```diff
@@ -42,26 +43,33 @@ data BuilderSt tag v x t = BuilderSt
     { signalBusWidth :: Int
     , availPorts :: [SignalTag]
     , puProtos :: [(tag, PU v x t)]
-    , netEnv :: UnitEnv (BusNetwork tag v x t)
     }
 
+netIOPorts puProtos =
+    let setOfPorts ports = foldl (flip $ S.union . ports . snd) S.empty puProtos
+     in BusNetworkIO
+            { extInputs = setOfPorts puInputPorts
+            , extOutputs = setOfPorts puOutputPorts
+            , extInOuts = setOfPorts puInOutPorts
+            }
+
+puEnv tag ports ioports =
```

Comment:
A few functions up you name imports as ports. So it will be better if you rename ports to ctrl ports or signals.